File: 2021/src/day02/day02.swift
import Foundation

enum Day2 {
  static func parse(_ contents: String) -> [[Substring]] {
    return contents.split(separator: "\n").map { $0.split(separator: " ") }
  }

  static func part1(input: String) -> Int {
    let data = parse(input)
    var horizontal = 0, depth = 0

    data.forEach { value in
      let dir = value[0]
      let amount = Int(value[1]) ?? -1

      switch dir {
      case "forward":
        horizontal += amount
      case "down":
        depth += amount
      case "up":
        depth -= amount
      default:
        break
      }
    }

    return horizontal * depth
  }

  static func part2(input: String) -> Int {
    let data = parse(input)
    var aim = 0, horizontal = 0, depth = 0

    data.forEach { value in
      let dir = value[0]
      let amount = Int(value[1]) ?? -1

      switch dir {
      case "forward":
        horizontal += amount
        depth += amount * aim
      case "down":
        aim += amount
      case "up":
        aim -= amount
      default:
        break
      }
    }

    return horizontal * depth
  }
}

File: 2021/src/day02/day02.js
const parseInput = (raw) =>
  raw
    .trim()
    .split("\n")
    .map((v) => {
      const [direction, amount] = v.split(" ");
      return { direction, amount: +amount };
    });

export const part1 = (raw) => {
  const input = parseInput(raw);
  const position = [0, 0]; // horizontal, depth

  input.forEach(({ direction, amount }) => {
    switch (direction) {
      case "forward":
        position[0] += amount;
        break;
      case "up":
        amount = -amount;
      case "down":
        position[1] += amount;
        break;
    }
  });

  return position[0] * position[1];
};

export const part2 = (raw) => {
  const input = parseInput(raw);
  const position = [0, 0, 0]; // horizontal, depth, aim

  input.forEach(({ direction, amount }) => {
    switch (direction) {
      case "forward":
        position[0] += amount;
        position[1] += amount * position[2];
        break;
      case "up":
        amount = -amount;
      case "down":
        position[2] += amount;
        break;
    }
  });

  return position[0] * position[1];
};
